java/ql/lib/semmle/code/java/ControlFlowGraph.qll

```diff
@@ -867,24 +948,46 @@ private module ControlFlowGraphImpl {
       // any other abnormal completion is propagated
       last(switch.getAStmt(), last, completion) and
       not completion instanceof YieldCompletion and
-      completion != NormalCompletion()
+      not completion instanceof NormalOrBooleanCompletion
     )
     or
-    // the last node in a case rule is the last node in the right-hand side
-    // if the rhs is a statement we wrap the completion as a break
-    exists(Completion caseCompletion |
-      last(n.(SwitchCase).getRuleStatement(), last, caseCompletion) and
+    // If a case rule right-hand-side completes then the switch breaks or yields, depending
+    // on whether this is a switch expression or statement. If it completes abruptly then the
+    // switch completes the same way.
+    exists(Completion caseCompletion, SwitchCase case |
+      case = n and
+      (
+        last(case.getRuleStatement(), last, caseCompletion)
+        or
+        last(case.getRuleExpression(), last, caseCompletion)
+      )
+    |
       if caseCompletion instanceof NormalOrBooleanCompletion
-      then completion = anonymousBreakCompletion()
+      then
+        case.getParent() instanceof SwitchStmt and completion = anonymousBreakCompletion()
+        or
+        case.getParent() instanceof SwitchExpr and completion = YieldCompletion(caseCompletion)
       else completion = caseCompletion
     )
     or
-    // ...and if the rhs is an expression we wrap the completion as a yield
-    exists(Completion caseCompletion |
-      last(n.(SwitchCase).getRuleExpression(), last, caseCompletion) and
-      if caseCompletion instanceof NormalOrBooleanCompletion
-      then completion = YieldCompletion(caseCompletion)
-      else completion = caseCompletion
+    // A pattern case statement can complete:
+    // * On failure of its type test (boolean false)
+    // * On failure of its guard test if any (boolean false)
+    // * On any abrupt completion of its guard
+    // * On completion of its variable declarations, if it is not a rule and has no guard (normal completion)
+    // * On success of its guard test, if it is not a rule (boolean true)
+    // (the latter two cases are accounted for by lastPatternCaseMatchingOp)
+    exists(PatternCase pc | n = pc |
+      last = pc and completion = basicBooleanCompletion(false)
+      or
+      last(pc.getGuard(), last, completion) and
+      (
+        completion = BooleanCompletion(false, _) or
+        not completion instanceof NormalOrBooleanCompletion
```

Comment:
This line is superfluous. There's a catch-all propagation of these completions at the very top of the definition of `last`.